### ISSUE TITLE
Kernel panic in bento/ubuntu-22.04 for libvirt

### ISSUE THREAD

**Issue by AndrewKvalheim** [chef/bento#1558]
## Version

- [bento/ubuntu-22.04](https://app.vagrantup.com/bento/boxes/ubuntu-22.04) 202309.08.0

## Environment

- Vagrant 2.4.1
- libvirt 9.9.0
- QEMU 8.1.5
- Linux 6.1.81 x86_64
- NixOS 23.11.5328.51063ed4f234

## Scenario

Run an Ubuntu LTS release in Vagrant

## Steps to Reproduce

```console
$ vagrant init 'bento/ubuntu-22.04'
$ vagrant up
```

## Expected Result

Provisioned Vagrant machine

## Actual Result

Kernel panic at boot:

```console
$ vagrant init 'bento/ubuntu-22.04'
A `Vagrantfile` has been placed in this directory. You are now
ready to `vagrant up` your first virtual environment! Please read
the comments in the Vagrantfile as well as documentation on
`vagrantup.com` for more information on using Vagrant.
$ vagrant up
Bringing machine 'default' up with 'libvirt' provider...
==> default: Checking if box 'bento/ubuntu-22.04' version '202309.08.0' is up to date...
==> default: Creating image (snapshot of base box volume).
==> default: Creating domain with the following settings...
==> default:  -- Name:              bento_default
==> default:  -- Description:       Source: /home/user/example/Vagrantfile
==> default:  -- Domain type:       qemu
==> default:  -- Cpus:              1
==> default:  -- Feature:           acpi
==> default:  -- Feature:           apic
==> default:  -- Feature:           pae
==> default:  -- Clock offset:      utc
==> default:  -- Memory:            512M
==> default:  -- Base box:          bento/ubuntu-22.04
==> default:  -- Storage pool:      default
==> default:  -- Image(vda):        /var/lib/libvirt/images/bento_default.img, virtio, 64G
==> default:  -- Disk driver opts:  cache='default'
==> default:  -- Graphics Type:     vnc
==> default:  -- Video Type:        cirrus
==> default:  -- Video VRAM:        16384
==> default:  -- Video 3D accel:    false
==> default:  -- Keymap:            en-us
==> default:  -- TPM Backend:       passthrough
==> default:  -- INPUT:             type=mouse, bus=ps2
==> default: Creating shared folders metadata...
==> default: Starting domain.
==> default: Domain launching with graphics connection settings...
==> default:  -- Graphics Port:      5900
==> default:  -- Graphics IP:        127.0.0.1
==> default:  -- Graphics Password:  Not defined
==> default:  -- Graphics Websocket: 5700
==> default: Waiting for domain to get an IP address...
```

![Kernel panic - not syncing: System is deadlocked on memory](https://github.com/chef/bento/assets/1844746/73b5f81f-19e6-4102-8cc0-e8faf42facd8)

**Comment by skyscooby** [comment]
Did you try increasing the memory allocated to the VM? It looks like 512M is not enough for Ubuntu 22.04?

**Comment by skyscooby** [comment]
> also have this problem, with enough resource:

This is probably not the same problem actually.. the libvirt image got busted with the latest release.. reported already as #1604.

You can pin back your box image to 202309.08.0 and it should boot fine with enough resources.

vm.box_version = 202309.08.0

**Comment by AndrewKvalheim** [comment]
> Did you try increasing the memory allocated to the VM? It looks like 512M is not enough for Ubuntu 22.04?

Thanks, looks like it:

| Box version | Default behavior | With override to 536 MB |
|-------------|------------------|-------------------------|
| 202304.25.0 | kernel panic     | success                 |
| 202309.08.0 | kernel panic     | success                 |
| 202502.21.0 | kernel panic     | success                 |

Can the box be configured with an appropriate default?

**Comment by skyscooby** [comment]
> Can the box be configured with an appropriate default?

I wish I knew how.. from what I've read there is a Vagrantfile that is generated with `vagrant package` step that is included with the box itself and loaded first and that seems to be where one would place a default value for memory...  

This project used to be a lot more straightforward with specific .json files for each build but now it's all gone into ruby and it seems like all the packer files are being dynamically generated.. I don't have time to figure it out nor does there seem to be any CI logs published so I guess they are internal to Progress on a private Github instance maybe?.. possibly just built locally by the current maintainer?.. anyhow there isn't even tags in this repository representing the builds since 2023 so I really think this whole project is likely becoming stale sadly.

**Comment by Stromweld** [comment]
The project has always been a ruby project, but the bento ruby gem is simply a wrapper for packer to loop through all the OS configs and build images for each OS type/version. Packer deprecated the .json config files in favor of their more universal HCL2 language and all the templates in the $BENTO_DIR/packer_templates and in $BENTO_DIR/os_pkrvars folders have been upgraded to hcl2 standards now.

Everything is public here, this repo is using Github Actions CI system. We don't have any internal CI system for this project. But due to limitations the CI system is still a work in progess. In the past all the releases have been built locally on mine or my predecessors mac systems and uploaded. Past maintainers have left the company and I'm the only maintainer now for past 2 years doing it on the side when I have extra time or on my off time.

In 2023 I stopped releasing based on dated versions. The decision to switch to semantic versioning was for the bento gem itself as it better communicates changes. This project should of really been 2 separate repos. 1 for the packer templates and 1 for the bento app itself. The boxes are still published based on dated versioning scheme as that best communicates to users age of the image in use and sets an expected patch level based on date of release.

We strive for out of box support for parallels, vmware-fusion/workstation, virtualbox for arm64 and x86_64. Libvirt, qemu, hyper-v have always been best effort I'm happy to review PR submissions to fix issues you find and can help contribute fixes for. qemu i've found to be a bit more tricky due to the lack of knowledge on my part and it's plethora of configuration options for a system. Especially for EFI based systems it seems to be even harder due to packaging and file location differences in macs and linux distros for the efi var files.

**Comment by skyscooby** [comment]
@Stromweld it's much appreciated the effort you are able to put in..  having only one maintainer on an open source project also presents significant risks to it's long term existence, I guess that's a budget thing or lack of interest / expertise in the community? But this is what has happened with the "roboxes" stream of images.. the single maintainer seems to have just gone AWOL and stopped responding to the community.. 

It seems like vagrant ecosystem is stale in general no idea what IBM has in store for it.. maybe it's partially due to Kubernetes eating the DevOps world..  but please know that there are still plenty of use cases for these images and it would be really nice to get it setup to be maintainable long term.

---

To close off this issue we need to increase it because the 512MB default is non functional in some OS's.. I've seen some previous commits in this project increasing the build instances to 1024+.. but this doesn't seem to persist with libvirt images.

Vagrant itself sets the default to 2048 but only for 3 of the providers:
https://github.com/hashicorp/vagrant/blob/e98ee62126c3f610e4970954768a1674db0a5889/Vagrantfile#L13

The libvirt plugin is still using 512MB which is where I believe these boxes are getting their config from.
https://github.com/vagrant-libvirt/vagrant-libvirt/blob/a94ce0d7b6c90129a38435698ca97b364130313d/lib/vagrant-libvirt/config.rb#L972


One way to fix this would be to use the `Vagrantfile` that is packed into the box itself to set an amount of memory to a value that works for the image. I'm not sure if any of your existing images do this? That's the easiest fix if possible.. otherwise we are gonna have to try and convince the libvirt plugin maintainers to change the default to a higher value...

**Comment by Stromweld** [comment]
Can you guys try this box https://portal.cloud.hashicorp.com/vagrant/discover/utm/bookworm with libvirt. It's built with UTM but that's just a GUI around qemu much like libvirt. I'm curious if I switch to that if that may help build libvirt compatible boxes.

**Comment by skyscooby** [comment]
> It's built with UTM but that's just a GUI around qemu

Most people using the libvirt image build would be using KVM not QEMU.. QEMU is slow software emulation while KVM is like Hyper-V .. it's an API implemented in the kernel and is just about as fast as native code... 

I'm not sure how I'd test this image on Linux anyway since it's declared as a UTM provider in vagrant cloud which means I'd need to install UTM software / plugin.. and UTM is only available as a Mac .dmg file...

**Comment by Stromweld** [comment]
qemu also uses kvm. In the past the boxes have only been built with qemu builder and uploaded to libvirt provider like as seen here https://github.com/chef/bento/blob/bento_old_json_templates/packer_templates/ubuntu/ubuntu-22.04-amd64.json#L174

**Comment by skyscooby** [comment]
> qemu also uses kvm.

yes I believe they both use the qcows2 disk format are are generally compatible but it's software emulation vs hardware virtualization.. it's also not the default for the libvirt provider you have to explicitly say `driver = "kvm"` if you want the performance boost...  Something changed between the build of 2023 and 2025 though.. as I use the identical Vagrantfile only changing the version and I go from working to can't boot at all.
